File: sdk/reproit-linux/reproit_linux/causal.py
"""Fail-closed causal capture/replay for Python's process-wide urllib client."""

import io
import json
import os
import re
import threading
import urllib.parse
import urllib.request

_SECRET = re.compile(
    r"password|passwd|secret|token|authorization|cookie|email|phone|"
    r"api[-_. ]?key|publishable[-_. ]?key|private[-_. ]?key|"
    r"access[-_. ]?key|signing[-_. ]?key",
    re.I,
)
_LOCK = threading.RLock()


def _redact(value):
    if isinstance(value, list):
        return [_redact(item) for item in value]
    if isinstance(value, dict):
        return {
            str(key): (
                "<reproit:string:length=%d>" % len(child)
                if _SECRET.search(str(key)) and isinstance(child, str)
                else "<reproit:secret>"
                if _SECRET.search(str(key))
                else _redact(child)
            )
            for key, child in sorted(value.items(), key=lambda item: str(item[0]))
        }
    return value


def _url(raw):
    parsed = urllib.parse.urlsplit(raw)
    query = urllib.parse.urlencode(
        sorted(urllib.parse.parse_qsl(parsed.query, keep_blank_values=True))
    )
    return urllib.parse.urlunsplit(
        (parsed.scheme.lower(), parsed.netloc.lower(), parsed.path, query, parsed.fragment)
    )


def _body(raw, headers):
    if not raw:
        return None
    content_type = next(
        (value for key, value in headers.items() if key.lower() == "content-type"), ""
    )
    if "json" in content_type.lower():
        try:
            return _redact(json.loads(raw.decode("utf-8")))
        except Exception:
            return "<reproit:invalid-json>"
    return "<reproit:body:length=%d>" % len(raw)


class _Response:
    def __init__(self, raw, status, headers, url):
        self._stream = io.BytesIO(raw)
        self.status = status
        self.code = status
        self.headers = headers
        self.url = url

    def read(self, amt=-1):
        return self._stream.read(amt)

    def getcode(self):
        return self.status

    def geturl(self):
        return self.url

    def close(self):
        self._stream.close()

    def __enter__(self):
        return self

    def __exit__(self, *_args):
        self.close()


def _action_index():
    try:
        with open(os.environ.get("REPROIT_ACTION_FILE", ""), encoding="utf-8") as handle:
            return int(handle.read().strip() or "0")
    except Exception:
        return 0


def _write_capabilities(path):
    if not path:
        return
    try:
        with open(path, encoding="utf-8") as handle:
            value = json.load(handle)
    except Exception:
        value = {}
    value["http"] = {"status": "captured", "detail": "Python urllib process hook"}
    value["http_replay"] = {
        "status": "captured",
        "detail": "Python urllib fail-closed replay",
    }
    try:
        with open(path, "w", encoding="utf-8") as handle:
            json.dump(value, handle, sort_keys=True, separators=(",", ":"))
    except Exception:
        pass


def install_causal_urllib(exclude_prefix=None):
    """Install once during a Reproit run and return an idempotent restore callback."""
    network = os.environ.get("REPROIT_NETWORK_FILE")
    capsule_path = os.environ.get("REPROIT_CAPSULE")
    if not network and not capsule_path:
        return lambda: None
    original = urllib.request.urlopen
    capsule = None
    if capsule_path:
        try:
            with open(capsule_path, encoding="utf-8") as handle:
                capsule = json.load(handle)
        except Exception:
            capsule = {"exchanges": []}
    exchanges = (capsule or {}).get("exchanges", [])
    used = set()
    state = {"action": -1, "ordinal": 0}

    def wrapped(value, *args, **kwargs):
        request = value if isinstance(value, urllib.request.Request) else urllib.request.Request(value)
        url = request.full_url
        if exclude_prefix and url.startswith(exclude_prefix):
            return original(value, *args, **kwargs)
        action = _action_index()
        with _LOCK:
            if state["action"] != action:
                state.update(action=action, ordinal=0)
            ordinal = state["ordinal"]
            state["ordinal"] += 1
        method = request.get_method().upper()
        actor = os.environ.get("REPROIT_DEVICE", "a")
        if capsule is not None:
            with _LOCK:
                match = next(
                    (
                        (index, exchange)
                        for index, exchange in enumerate(exchanges)
                        if index not in used
                        and exchange.get("required")
                        and exchange.get("actor") == actor
                        and exchange.get("actionIndex", exchange.get("action_index")) == action
                        and str(exchange.get("method", "")).upper() == method
                        and _url(str(exchange.get("url", ""))) == _url(url)
                    ),
                    None,
                )
                if match:
                    used.add(match[0])
            if not match:
                raise RuntimeError("CAPSULE:MISS %s %s action=%d" % (method, url, action))
            exchange = match[1]
            body = exchange.get("responseBody", exchange.get("response_body", ""))
            raw = body.encode("utf-8") if isinstance(body, str) else json.dumps(body).encode("utf-8")
            return _Response(raw, int(exchange.get("status", 200)), exchange.get("responseHeaders", exchange.get("response_headers", {})), url)

        response = original(value, *args, **kwargs)
        raw = response.read()
        response_headers = dict(response.headers.items())
        request_headers = dict(request.header_items())
        safe = lambda headers: {
            key: "<reproit:secret>" if _SECRET.search(key) else val
            for key, val in headers.items()
        }
        exchange = {
            "id": "%s-%d-%d" % (actor, action, ordinal),
            "actor": actor,
            "actionIndex": action,
            "ordinal": ordinal,
            "protocol": urllib.parse.urlsplit(url).scheme,
            "method": method,
            "url": url,
            "requestHeaders": safe(request_headers),
            "requestBody": _body(request.data, request_headers),
            "status": response.status,
            "responseHeaders": safe(response_headers),
            "responseBody": _body(raw, response_headers),
            "required": True,
        }
        try:
            with open(network, "a", encoding="utf-8") as handle:
                handle.write(json.dumps(exchange, sort_keys=True, separators=(",", ":")) + "\n")
        except Exception:
            pass
        return _Response(raw, response.status, response_headers, url)

    urllib.request.urlopen = wrapped
    _write_capabilities(os.environ.get("REPROIT_CAPABILITIES_FILE"))
    return lambda: setattr(urllib.request, "urlopen", original)

File: sdk/reproit_flutter/lib/src/causal.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

final RegExp _secret = RegExp(
  r'password|passwd|secret|token|authorization|cookie|email|phone|'
  r'api[-_. ]?key|publishable[-_. ]?key|private[-_. ]?key|'
  r'access[-_. ]?key|signing[-_. ]?key',
  caseSensitive: false,
);

Object? redactCausal(Object? value) {
  if (value is List) return value.map(redactCausal).toList();
  if (value is Map) {
    final out = <String, Object?>{};
    final entries = value.entries.toList()
      ..sort((a, b) => a.key.toString().compareTo(b.key.toString()));
    for (final entry in entries) {
      final key = entry.key.toString();
      final child = entry.value;
      if (_secret.hasMatch(key)) {
        final type = child is String
            ? 'string:length=${child.runes.length}'
            : child.runtimeType.toString().toLowerCase();
        out[key] = '<reproit:$type>';
      } else {
        out[key] = redactCausal(child);
      }
    }
    return out;
  }
  return value;
}

Map<String, String> _headers(Map<String, String> headers) => {
      for (final entry
          in (headers.entries.toList()..sort((a, b) => a.key.compareTo(b.key))))
        entry.key:
            _secret.hasMatch(entry.key) ? '<reproit:secret>' : entry.value,
    };

String _canonicalUrl(Uri uri) {
  final pairs = <MapEntry<String, String>>[];
  uri.queryParametersAll.forEach((key, values) {
    for (final value in values) pairs.add(MapEntry(key, value));
  });
  pairs.sort((a, b) {
    final byKey = a.key.compareTo(b.key);
    return byKey != 0 ? byKey : a.value.compareTo(b.value);
  });
  final base = uri.replace(query: '').toString();
  if (pairs.isEmpty) return base;
  final query = pairs.map((entry) {
    final key = Uri.encodeQueryComponent(entry.key);
    final value = Uri.encodeQueryComponent(entry.value);
    return '$key=$value';
  }).join('&');
  return '$base?$query';
}

class ReproItCausalClient extends http.BaseClient {
  final http.Client _inner;
  final int Function() actionIndex;
  final String actor;
  final List<Map<String, dynamic>>? _replay;
  final Set<int> _used = {};
  int _priorAction = 0;
  int _ordinal = 0;

  ReproItCausalClient({
    required this.actionIndex,
    this.actor = 'a',
    http.Client? inner,
    List<Map<String, dynamic>>? replay,
  })  : _inner = inner ?? http.Client(),
        _replay = replay {
    print(
        'REPROIT:CAPABILITIES {"http":{"status":"captured"},"http_replay":{"status":"captured"}}');
  }

  factory ReproItCausalClient.fromEnvironment(
      {required int Function() actionIndex}) {
    const raw = String.fromEnvironment('REPROIT_CAPSULE_JSON');
    List<Map<String, dynamic>>? exchanges;
    if (raw.isNotEmpty) {
      final decoded = jsonDecode(raw) as Map<String, dynamic>;
      exchanges = (decoded['exchanges'] as List<dynamic>? ?? const [])
          .whereType<Map>()
          .map((value) => value.map((k, v) => MapEntry(k.toString(), v)))
          .toList();
    }
    return ReproItCausalClient(actionIndex: actionIndex, replay: exchanges);
  }

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) async {
    final action = actionIndex();
    if (action != _priorAction) {
      _priorAction = action;
      _ordinal = 0;
    }
    final ordinal = _ordinal++;
    final canonicalUrl = _canonicalUrl(request.url);
    if (_replay != null) {
      var match = -1;
      for (var i = 0; i < _replay!.length; i++) {
        final exchange = _replay![i];
        if (!_used.contains(i) &&
            exchange['required'] == true &&
            exchange['actor'] == actor &&
            exchange['actionIndex'] == action &&
            exchange['method'].toString().toUpperCase() ==
                request.method.toUpperCase() &&
            _canonicalUrl(Uri.parse(exchange['url'].toString())) ==
                canonicalUrl) {
          match = i;
          break;
        }
      }
      if (match < 0) {
        print('CAPSULE:MISS ${request.method} ${request.url} action=$action');
        throw StateError(
            'CAPSULE:MISS ${request.method} ${request.url} action=$action');
      }
      _used.add(match);
      final exchange = _replay![match];
      print('CAPSULE:HIT ${exchange['id']}');
      final body = exchange['responseBody'];
      final bytes = utf8.encode(body is String ? body : jsonEncode(body ?? ''));
      return http.StreamedResponse(
        http.ByteStream.fromBytes(bytes),
        exchange['status'] as int,
        headers: (exchange['responseHeaders'] as Map? ?? const {})
            .map((k, v) => MapEntry(k.toString(), v.toString())),
        request: request,
      );
    }

    Object? requestBody;
    if (request is http.Request && request.body.isNotEmpty) {
      try {
        requestBody = redactCausal(jsonDecode(request.body));
      } catch (_) {
        requestBody =
            '<reproit:body:length=${utf8.encode(request.body).length}>';
      }
    }
    final response = await _inner.send(request);
    final bytes = await response.stream.toBytes();
    Object? responseBody;
    if ((response.headers['content-type'] ?? '').contains('json')) {
      try {
        responseBody = redactCausal(jsonDecode(utf8.decode(bytes)));
      } catch (_) {
        responseBody = '<reproit:invalid-json>';
      }
    } else {
      responseBody = '<reproit:body:length=${bytes.length}>';
    }
    final exchange = <String, Object?>{
      'id': '$actor-$action-$ordinal',
      'actor': actor,
      'actionIndex': action,
      'ordinal': ordinal,
      'protocol': request.url.scheme,
      'method': request.method,
      'url': request.url.toString(),
      'requestHeaders': _headers(request.headers),
      if (requestBody != null) 'requestBody': requestBody,
      'status': response.statusCode,
      'responseHeaders': _headers(response.headers),
      'responseBody': responseBody,
      'required': true,
    };
    print('REPROIT:EXCHANGE ${jsonEncode(exchange)}');
    return http.StreamedResponse(
      http.ByteStream.fromBytes(bytes),
      response.statusCode,
      contentLength: bytes.length,
      request: request,
      headers: response.headers,
      isRedirect: response.isRedirect,
      persistentConnection: response.persistentConnection,
      reasonPhrase: response.reasonPhrase,
    );
  }

  @override
  void close() => _inner.close();
}
